sqlglotrs/src/tokenizer.rs

```diff
@@ -0,0 +1,642 @@
+        while self.size > 0 && !self.is_end {
+            self.start = self.current;
+            self.advance(1, false, false);
+
+            if self.current_char == '\0' {
+                break;
+            }
+
+            if !self.settings.white_space.contains_key(&self.current_char) {
+                if self.current_char.is_digit(10) {
+                    self.scan_number();
+                } else {
+                    match self.settings.identifiers.get(&self.current_char) {
+                        Some(identifier_end) => self.scan_identifier(&identifier_end.to_string()),
+                        None => self.scan_keyword(),
+                    }
+                }
+            }
+
+            match until_peek_char {
+                Some(c) => {
+                    if self.peek_char == c {
+                        break;
+                    }
+                }
+                None => (),
+            }
+        }
+        if !self.tokens.is_empty() && !self.comments.is_empty() {
+            self.tokens
+                .last_mut()
+                .unwrap()
+                .append_comments(&mut self.comments);
+        }
+    }
+
+    fn advance(&mut self, i: usize, alnum: bool, backtrack: bool) {
+        match self.settings.white_space.get(&self.current_char) {
+            Some(TokenType::BREAK) => {
+                self.column = 1;
+                self.line += 1;
+            }
+            _ => {
+                if !backtrack {
+                    self.column += i
+                } else {
+                    self.column -= i
+                }
+            }
+        }
+
+        if !backtrack {
+            self.current += i
+        } else {
+            self.current -= i
+        };
+
+        self.is_end = self.current >= self.size;
+        self.current_char = self.char_at(self.current - 1);
+        self.peek_char = if self.is_end {
+            '\0'
+        } else {
+            self.char_at(self.current)
+        };
+
+        if alnum && self.current_char.is_alphanumeric() {
+            while self.peek_char.is_alphanumeric() {
+                self.column += 1;
+                self.current += 1;
+                self.is_end = self.current >= self.size;
+                self.peek_char = if self.is_end {
+                    '\0'
+                } else {
+                    self.char_at(self.current)
+                };
+            }
+            self.current_char = self.char_at(self.current - 1);
+        }
+    }
+
+    fn peek(&self, i: usize) -> char {
+        let index = self.current + i;
+        if index < self.size {
+            self.char_at(index)
+        } else {
+            '\0'
+        }
+    }
+
+    fn chars(&self, size: usize) -> String {
+        let start = self.current - 1;
+        let end = start + size;
+        if end <= self.size {
+            self.sql[start..end].iter().collect()
+        } else {
+            String::from("")
```

Comment:
is this a constant? or is it being allocated every time?

Reply:
It's allocated on heap every time. The type system doesn't permit using a constant value here, since String type is mutable in rust.